homeassistant/components/enocean/binary_sensor.py

```diff
@@ -26,31 +36,90 @@
     dev_name = config.get(CONF_NAME)
     device_class = config.get(CONF_DEVICE_CLASS)
+    sensor = EnOceanBinarySensor(dev_id, dev_name, device_class)
+
+    entity_reg = await async_get_registry(hass)
+    configured_entities = async_entries_for_config_entry(
+        entity_reg, config_entry.entry_id
+    )
+
+    entities_to_migrate = hass.data[DATA_ENOCEAN].setdefault(
+        YAML_BINARY_SENSORS_TO_IMPORT, []
+    )
+    for entity in configured_entities:
+        if entity.unique_id == sensor.unique_id:
+            _LOGGER.warning(
+                "Yaml config for binary sensor %s ignored: entity exists in registry",
+                dev_name,
+            )
+            return
+
+    entities_to_migrate.append(EnOceanBinarySensor(dev_id, dev_name, device_class,))
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry, async_add_entities):
+    """Set up binary sensors for the ENOcean component."""
+
+    entities_to_add = []
+
+    # Load entities from registry
+    entities_registry = await entity_registry.async_get_registry(hass)
+    entities_from_config = async_entries_for_config_entry(
+        entities_registry, config_entry.entry_id
+    )
+    for current_entity in entities_from_config:
+        entities_to_add.append(
+            EnOceanBinarySensor(
+                string_to_enocean_id(current_entity.unique_id.split("-")[0]),
+                current_entity.name,
+                current_entity.device_class,
+            )
+        )
 
-    add_entities([EnOceanBinarySensor(dev_id, dev_name, device_class)])
+    # Import entities from yaml configuration
+    if YAML_BINARY_SENSORS_TO_IMPORT in hass.data[DATA_ENOCEAN]:
+        entities_to_migrate = hass.data[DATA_ENOCEAN][YAML_BINARY_SENSORS_TO_IMPORT]
+        for entity in entities_to_migrate:
+            _LOGGER.info("Importing yaml config for binary sensor %s", entity.dev_name)
+        entities_to_add.extend(entities_to_migrate)
+        hass.data[DATA_ENOCEAN].pop(YAML_BINARY_SENSORS_TO_IMPORT)
 
+    async_add_entities(entities_to_add, True)
 
-class EnOceanBinarySensor(enocean.EnOceanDevice, BinarySensorEntity):
+
+class EnOceanBinarySensor(EnOceanEntity, BinarySensorEntity):
     """Representation of EnOcean binary sensors such as wall switches.
 
     Supported EEPs (EnOcean Equipment Profiles):
     - F6-02-01 (Light and Blind Control - Application Style 2)
     - F6-02-02 (Light and Blind Control - Application Style 1)
     """
 
-    def __init__(self, dev_id, dev_name, device_class):
+    def __init__(self, dev_id: List[int], dev_name: str, device_class: str):
         """Initialize the EnOcean binary sensor."""
         super().__init__(dev_id, dev_name)
         self._device_class = device_class
         self.which = -1
         self.onoff = -1
 
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Return a unique ID for the device."""
+        return "{}-binary_sensor".format(enocean_id_to_string(self.dev_id))
```

Comment:
We don't need to add `binary_sensor` to the unique_id. The entity registry is aware of both the integration domain and the platform domain.

Reply:
That is cool, it does make things easier :-).
Change done.